File: crypto-curriculum/modules/10-java/src/io/learncrypto/Ciphers.java
package io.learncrypto;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import java.security.SecureRandom;
import java.util.Arrays;

/** 使用 12 字节 IV 的 AES-256-GCM 加解密。 */
public final class Ciphers {
    public static final int TAG_BITS = 128;
    private static final SecureRandom RNG = new SecureRandom();

    private Ciphers() {}

    public static final class GcmEnv {
        public final byte[] ct;
        public final byte[] nonce;
        public final byte[] tag;
        public GcmEnv(byte[] ct, byte[] nonce, byte[] tag) {
            this.ct = ct; this.nonce = nonce; this.tag = tag;
        }
    }

    public static GcmEnv aesGcmEncrypt(byte[] key, byte[] pt, byte[] aad) throws GeneralSecurityException {
        if (key.length != 32) throw new IllegalArgumentException("key must be 32 bytes");
        byte[] nonce = new byte[12];
        RNG.nextBytes(nonce);
        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        c.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"),
                new GCMParameterSpec(TAG_BITS, nonce));
        if (aad != null && aad.length > 0) c.updateAAD(aad);
        byte[] ct = c.doFinal(pt); // 追加 16 字节的 tag
        // 拆分 ct 与 tag。
        byte[] ctOnly = Arrays.copyOf(ct, ct.length - 16);
        byte[] tag    = Arrays.copyOfRange(ct, ct.length - 16, ct.length);
        return new GcmEnv(ctOnly, nonce, tag);
    }

    public static byte[] aesGcmDecrypt(byte[] key, GcmEnv env, byte[] aad) throws GeneralSecurityException {
        if (key.length != 32) throw new IllegalArgumentException("key must be 32 bytes");
        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"),
                new GCMParameterSpec(TAG_BITS, env.nonce));
        if (aad != null && aad.length > 0) c.updateAAD(aad);
        // 重新组合 ct||tag，因为 cipher.doFinal 需要传入合并后的密文。
        ByteBuffer buf = ByteBuffer.allocate(env.ct.length + env.tag.length);
        buf.put(env.ct); buf.put(env.tag);
        return c.doFinal(buf.array());
    }
}

File: crypto-curriculum/modules/13-python/src/learncrypto/ciphers.py
"""使用 pyca 的 cryptography.hazmat 实现 AES-256-GCM 的加解密往返。

运行方式::

    python -c "from learncrypto.ciphers import *; ..."
"""

from __future__ import annotations

import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def aes_gcm_encrypt(key: bytes, pt: bytes, aad: bytes | None = None) -> dict[str, bytes]:
    """使用 AES-256-GCM 加密 `pt`。返回 {ciphertext, nonce, tag}。

    `cryptography` 的 AESGCM.Seal 返回的是 ct||tag 拼接后的结果；我们将其拆分。
    """
    if len(key) != 32:
        raise ValueError("key must be 32 bytes")
    nonce = os.urandom(12)
    aes = AESGCM(key)
    sealed = aes.encrypt(nonce, pt, aad)
    ct, tag = sealed[:-16], sealed[-16:]
    return {"ciphertext": ct, "nonce": nonce, "tag": tag}


def aes_gcm_decrypt(key: bytes, env: dict[str, bytes], aad: bytes | None = None) -> bytes:
    if len(key) != 32:
        raise ValueError("key must be 32 bytes")
    aes = AESGCM(key)
    return aes.decrypt(env["nonce"], env["ciphertext"] + env["tag"], aad)
